drivers/nvme.cc

```diff
@@ -0,0 +1,693 @@
+    blk_ioctl,
+    no_devctl,
+    multiplex_strategy,
+};
+
+struct driver nvme_driver = {
+    "nvme",
+    &nvme_devops,
+    sizeof(struct nvme_priv),
+};
+
+int nvme::_instance = 0;
+
+extern std::vector<sched::cpu*> sched::cpus;
+
+nvme::nvme(pci::device &dev)
+     : _dev(dev)
+     , _msi(&dev)
+{
+    parse_pci_config();
+    u16 command = dev.get_command();
+    command |= 0x4 | 0x2 | 0x400;
+    dev.set_command(command);
+
+    _id = _instance++;
+    
+    _doorbellstride = 1 << (2 + _control_reg->cap.dstrd);
+    
+    wait_for_controller_ready_change(1);
+    disable_controller();
+
+    init_controller_config();
+
+    create_admin_queue();
+    
+    enable_controller();
+
+    identify_controller();
+
+    if(NVME_CHECK_FOR_ADDITIONAL_NAMESPACES) {
+        identify_active_namespaces(1);
+    } else {
+        identify_namespace(1);
+    }
+
+    if(_identify_controller->vwc & 0x1 && NVME_VWC_ENABLED) {
+        auto cmd = alloc_cmd();
+        cmd->set_features.common.opc = NVME_ACMD_SET_FEATURES;
+        cmd->set_features.fid = NVME_FEATURE_WRITE_CACHE;
+        cmd->set_features.val = 1;
+        auto res = _admin_queue->submit_and_return_on_completion(std::move(cmd));
+        trace_nvme_vwc_enabled(res->sc,res->sct);
+    }
+
+    if(NVME_QUEUE_PER_CPU_ENABLED) {
+        u16 num = sched::cpus.size();
+        u16 ret;
+        set_number_of_queues(num, &ret);   
+        create_io_queues_foreach_cpu();
+    }else {
+        u16 ret;
+        set_number_of_queues(1, &ret);
+        assert(ret>=1);
+        create_io_queue();
+    }
+
+    if(_identify_controller->vid != QEMU_VID) {
+        set_interrupt_coalescing(20,2);
+    }
+
+    struct nvme_priv* prv;
+    struct device *osv_dev;
+    
+    debugf("nvme: %s\n", _identify_controller->sn);
+    
+    for(const auto& ns : _ns_data) {
+        std::string dev_name;
+        if(ns.first == 1 && _id == 0) {
+            dev_name = "vblk";
+            dev_name += std::to_string(_id);
+        } else {
+            dev_name = "nvme";
+            dev_name += std::to_string(_id) + "n";
+            dev_name += std::to_string(ns.first);
+        }
+        off_t size = ((off_t) ns.second->blockcount) << ns.second->blockshift;
+        
+        debugf("nvme: Add namespace %d of nvme device %d as %s, devsize=%lld\n", ns.first, _id, dev_name.c_str(), size);
+
+        osv_dev = device_create(&nvme_driver,dev_name.c_str(), D_BLK);
+        prv = reinterpret_cast<struct nvme_priv*>(osv_dev->private_data);
+        prv->strategy = nvme_strategy;
+        prv->drv = this;
+        prv->nsid = ns.first;
+        osv_dev->size = size;
+        /*
+        * IO size greater than 4096 << 9 would mean we need 
+        * more than 1 page for the prplist which is not implemented
+        */
+        osv_dev->max_io_size = 4096 << ((9 < _identify_controller->mdts)? 9 : _identify_controller->mdts );
+
+        #if CONF_drivers_io_test
```

Comment:
What is the motivation behind running this test during the driver initialization stage?